### ISSUE TITLE
torch.dot under vmap lowers to unfused extern_kernels.bmm — pointwise mul+sum is 25-30% faster for small vectors

### ISSUE THREAD

**Issue by vmoens** [pytorch/pytorch#176706]
## 🐛 Describe the bug

When `torch.dot(a, b)` is used inside `torch.vmap` and compiled with `torch.compile`, it gets batched into `torch.bmm` and lowered to `extern_kernels.bmm` (cuBLAS). For small vectors (3-4 elements, typical in physics simulations), this is significantly slower than the equivalent `(a * b).sum(-1)` because:

1. **cuBLAS dispatch overhead** dominates for tiny matrices like `(B, 1, 3) @ (B, 3, 1)`
2. **Fusion barrier** — `extern_kernels.bmm` is opaque to Inductor, so surrounding pointwise ops cannot be fused into the same Triton kernel
3. **Kernel count explosion** — each dot product becomes a standalone kernel launch

## Benchmark evidence

In [mujoco-torch](https://github.com/vmoens/mujoco-torch) (a PyTorch physics simulator), replacing every `torch.dot` call with `(a * b).sum(-1)` yielded **+25–30% throughput improvement** in `torch.compile` mode (H200, float64, B=32768). The codebase had ~40 `torch.dot` call sites on 3-element and 4-element vectors (positions, quaternion components).

Before the change, the compiled Inductor output for a single model contained **83 `extern_kernels.bmm` calls**, most operating on shapes like `(32768, 1, 3) @ (32768, 3, 1)`. After the change, these were all fused into neighboring Triton kernels.

## Minimal repro

```python
import torch

def dot_based(a, b):
    # 10 chained dot products (simulating a physics pipeline)
    s = torch.dot(a, b)
    for _ in range(9):
        s = s + torch.dot(a, b)
    return s

def pointwise_based(a, b):
    s = (a * b).sum(-1)
    for _ in range(9):
        s = s + (a * b).sum(-1)
    return s

B = 32768
a = torch.randn(B, 3, dtype=torch.float64, device="cuda")
b = torch.randn(B, 3, dtype=torch.float64, device="cuda")

vmap_dot = torch.vmap(dot_based)
vmap_pw  = torch.vmap(pointwise_based)

compiled_dot = torch.compile(vmap_dot, fullgraph=True)
compiled_pw  = torch.compile(vmap_pw, fullgraph=True)

# Warmup
for _ in range(3):
    compiled_dot(a, b)
    compiled_pw(a, b)
torch.cuda.synchronize()

import time

torch.cuda.synchronize()
t0 = time.perf_counter()
for _ in range(1000):
    compiled_dot(a, b)
torch.cuda.synchronize()
t_dot = time.perf_counter() - t0

torch.cuda.synchronize()
t0 = time.perf_counter()
for _ in range(1000):
    compiled_pw(a, b)
torch.cuda.synchronize()
t_pw = time.perf_counter() - t0

print(f"torch.dot:      {t_dot:.3f}s")
print(f"mul+sum:         {t_pw:.3f}s")
print(f"speedup:         {t_dot / t_pw:.2f}x")
```

## Inductor output comparison

**With `torch.dot`** (via vmap → bmm):
```python
extern_kernels.bmm(reinterpret_tensor(buf42, (32768, 1, 3), ...), 
                    reinterpret_tensor(buf43, (32768, 3, 1), ...), 
                    out=buf44)
```
Each call is a standalone cuBLAS invocation that breaks the fusion chain.

**With `(a * b).sum(-1)`:**
The multiply and sum are pointwise/reduction ops that Inductor fuses into the surrounding Triton kernel — no separate kernel launch.

## Suggested fix

Several possible approaches (not mutually exclusive):

1. **`vmap` batching rule for `dot`**: when the contraction dimension is small (e.g. ≤ 8), emit `(a * b).sum(-1)` instead of `bmm`
2. **Inductor lowering**: recognize tiny `bmm` (inner dim ≤ threshold) and lower to pointwise multiply + reduce instead of cuBLAS
3. **Inductor fusion pass**: allow small `extern_kernels.bmm` to be inlined as fused Triton ops

Option 2 seems most general — it would benefit any code path that produces small bmm, not just vmap'd dot products.

## Versions

- PyTorch: built from source (2.12.0a0+git8a4793d)
- GPU: NVIDIA H200
- CUDA: 12.x
- OS: Linux
